### PR TITLE
Run state accumalator pattern

### PR DESCRIPTION
**Pull request by xorima**
# Description

Describe what this change achieves

## Issues Resolved

List any existing issues this PR resolves

## Check List

- [ ] All tests pass. See TESTING.md for details.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by xorima** [comment]
Changes made:

Changed:
ldap servers host from name property to required property 
ldap group mappings group_dn from name property to required property 
added instance_name to above resources as name property
changed ldap config template from @grafana['ldap'] to @ldap
changes to service resource

Removed:
removed config directory from all config resource
removed config file from all config resource
removed cookbook from all config resource
removed source from all config resource

Added:
Sensitive on config_writer

ideas to think about:
Make changes to the s3 storage backend to sit in a better namespace, or make generic
all word arrays validated should become symbols because they look cleaner to me ...
We may also want to increase spec test coverage

**Review by JohnRoesler — APPROVED** [review]
looks good to me - a lot of the same changes, so while many lines, not too bad on the logic

File: documentation/grafana_config_database.md [hunk 1]
```diff
@@ -29,11 +29,7 @@ Introduced: v4.0.0
...
 | `server_cert_name`        |  String       |                             | The path to the client cert. Only if server requires client authentication|
-| `path`                    |  String       | `grafana.db`                | Only applicable for sqlite3 database. The file path where the database will be stored.|
-| `conf_directory`          | String        | `/etc/grafana`              | The directory where the Grafana configuration resides                     | Valid directory
-| `config_file`             | String        | `/etc/grafana/grafana.ini`  | The Grafana configuration file                                            | Valid file path
-| `cookbook`                | String        | `grafana`                   | Which cookbook to look in for the template                                |
-| `source`                  | String        | `grafana.ini.erb`           | Name of the template                                                      |
+| `path`                    |  String       | `grafana.db`                | Only applicable for sqlite3 database. The file path where the database will be|
```
Comment (axelrtgs) [inline documentation/grafana_config_database.md:32]: Why remove the word stored from the line. the sentence IMO was clearer with the word stored.

File: resources/config_users.rb [hunk 2]
```diff
@@ -37,40 +37,29 @@
 property  :source,                    String,         default: 'grafana.ini.erb'
 
 action :install do
-  with_run_context :root do
-    edit_resource(:template, new_resource.config_file) do |new_resource|
-      node.run_state['grafana'] ||= { 'conf_template_source' => {}, 'conf_cookbook' => {} }
-      source new_resource.source
```
Comment (axelrtgs) [inline resources/config_users.rb:43]: Confi_file, Cookbook and source properties above no longer needed

File: test/fixtures/cookbooks/test/recipes/install.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 service 'grafana-server' do
   action [:enable, :start]
-  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :delayed
+  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :immediately
```
Comment (axelrtgs) [inline test/fixtures/cookbooks/test/recipes/install.rb:5]: You recommend in the upgrading guide to remove this resource, why keep it and just change it to immediately.
I can understand keeping it around for the enable, start to make sure at the end of the run the service is started so grafana comes up.

File: test/fixtures/cookbooks/test/recipes/ldap.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 service 'grafana-server' do
   action [:enable, :start]
-  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :delayed
+  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :immediately
```
Comment (axelrtgs) [inline test/fixtures/cookbooks/test/recipes/ldap.rb:5]: Same as the install comment

File: test/fixtures/cookbooks/test/recipes/plugins.rb [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 service 'grafana-server' do
   action [:enable, :start]
-  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :delayed
+  subscribes :restart, ['template[/etc/grafana/grafana.ini]', 'template[/etc/grafana/ldap.toml]'], :immediately
```
Comment (axelrtgs) [inline test/fixtures/cookbooks/test/recipes/plugins.rb:5]: Same as install comment

**Review by axelrtgs — APPROVED** [review]

File: documentation/grafana_config_database.md (hunk 1 above)
Comment (axelrtgs) [reply documentation/grafana_config_database.md:32]: you could also change it to `The file path where to store the database`

**Pull request merged**